File: sdk/typescript/src/wire.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";
import * as wire from "./wire.js";
import type { Account, Transaction } from "./types.js";

test("a negative int64 is ten bytes", () => {
  assert.deepEqual(wire.encodePosting({ dim: "1", amount: "-1" }), [0x08, 0x01, 0x10, ...Array(9).fill(0xff), 0x01]);
});

test("absent optionals are not on the wire and a present zero is", () => {
  assert.deepEqual(wire.encodePosting({ dim: "0", amount: "0" }), []);
  assert.deepEqual(wire.encodePosting({ dim: "0", amount: "0", quantity: "0" }), [0x28, 0x00]);
});

test("a transaction round trips", () => {
  const t: Transaction = {
    name: "books/b/transactions/t",
    postings: [
      { dim: "2", amount: "-125000", currencyCode: "USD" },
      { dim: "1", amount: "125000", currencyCode: "USD", instrument: "AAPL", quantity: "10" },
    ],
    controlPlaneHash: "abc",
  };
  assert.deepEqual(wire.decodeTransaction(Uint8Array.from(wire.encodeTransaction(t))), t);
});

test("a create request nests the transaction", () => {
  const t: Transaction = { postings: [{ dim: "1", amount: "5" }] };
  const seen = new Map<number, Uint8Array>();
  for (const f of wire.fields(wire.encodeCreateTransactionRequest("books/b", t, "trade-1"))) seen.set(f.num, f.bytes);
  assert.equal(new TextDecoder().decode(seen.get(1)), "books/b");
  assert.deepEqual(wire.decodeTransaction(seen.get(2)!), t);
  assert.equal(new TextDecoder().decode(seen.get(3)), "trade-1");
});

test("accounts carry enums as numbers", () => {
  const a: Account = { name: "books/b/accounts/20", dim: "20", displayName: "Capital", accountType: "ACCOUNT_TYPE_EQUITY", normalSide: "SIDE_CREDIT" };
  const b = wire.encodeAccount(a);
  assert.ok(b.join(",").includes("32,3"), "field 4 = 3 (EQUITY)");
  assert.ok(b.join(",").includes("40,2"), "field 5 = 2 (CREDIT)");
  assert.deepEqual(wire.decodeAccount(Uint8Array.from(b)), a);
});

test("unknown fields are skipped and truncation is an error", () => {
  const extra = [0x48, 0x07, 0x51, ...Array(8).fill(0), 0x5a, 0x02, 0x68, 0x69];
  assert.deepEqual(wire.decodePosting(Uint8Array.from([...wire.encodePosting({ dim: "3", amount: "4" }), ...extra])), { dim: "3", amount: "4" });
  assert.throws(() => [...wire.fields(Uint8Array.from([0x12, 0x05, 0x61, 0x62]))]);
});

File: sdk/python/tests/wire_test.py
"""The wire codec: known bytes for known values, and round trips."""

import unittest

from ratio import _wire
from ratio.types import Account, AccountType, Posting, Side, Transaction


class WireTest(unittest.TestCase):
    def test_negative_int64_is_ten_bytes(self):
        # tag 0x08 (field 1, varint), 1; tag 0x10 (field 2, varint), -1 as 2^64-1.
        self.assertEqual(_wire.posting(Posting(1, -1)), b"\x08\x01\x10" + b"\xff" * 9 + b"\x01")

    def test_absent_optionals_are_not_on_the_wire_and_present_zero_is(self):
        self.assertEqual(_wire.posting(Posting(0, 0)), b"")
        self.assertEqual(_wire.posting(Posting(0, 0, quantity=0)), b"\x28\x00")

    def test_a_transaction_round_trips(self):
        t = Transaction(
            name="books/b/transactions/t",
            postings=(Posting(2, -125_000, "USD"), Posting(1, 125_000, "USD", "AAPL", 10)),
            control_plane_hash="abc",
        )
        self.assertEqual(_wire.decode_transaction(_wire.transaction(t)), t)

    def test_create_request_nests_the_transaction(self):
        t = Transaction(postings=(Posting(1, 5),))
        b = _wire.create_transaction_request("books/b", t, "trade-1")
        seen = {num: v for num, _, v in _wire.fields(b)}
        self.assertEqual(seen[1], b"books/b")
        self.assertEqual(_wire.decode_transaction(seen[2]), t)
        self.assertEqual(seen[3], b"trade-1")

    def test_accounts_carry_enums_as_numbers(self):
        a = Account("books/b/accounts/20", 20, "Capital", AccountType.EQUITY, Side.CREDIT)
        b = _wire.account(a)
        self.assertIn(b"\x20\x03", b)  # field 4 = 3 (EQUITY)
        self.assertIn(b"\x28\x02", b)  # field 5 = 2 (CREDIT)
        self.assertEqual(_wire.decode_account(b), a)

    def test_unknown_fields_are_skipped(self):
        # field 9 varint, field 10 fixed64, field 11 bytes — none in Posting.
        extra = b"\x48\x07" + b"\x51" + b"\x00" * 8 + b"\x5a\x02hi"
        self.assertEqual(_wire.decode_posting(_wire.posting(Posting(3, 4)) + extra), Posting(3, 4))

    def test_truncation_is_an_error(self):
        with self.assertRaises(ValueError):
            list(_wire.fields(b"\x12\x05ab"))


if __name__ == "__main__":
    unittest.main()
